torch/utils/data/communication/map.py

```diff
@@ -0,0 +1,143 @@
+                    NonBlockingMap.not_available_hook()
+
+    def nonblocking_len(self):
+        raise NotImplementedError(
+            "nonblocking_len is not implemented for %s" % self.__class__)
+
+    def nonblocking_getitem(self, index):
+        raise NotImplementedError(
+            "nonblocking_getitem is not implemented for %s" % self.__class__)
+
+    @staticmethod
+    def register_not_available_hook(hook_function):
+        NonBlockingMap.not_available_hook = hook_function
+
+
+def EnsureNonBlockingMapDataPipe(validated_datapipe):
+    if not isinstance(validated_datapipe, MapDataPipe):
+        raise Exception(f'Not Map DataPipe - got {validated_datapipe.__class__}')
+    if isinstance(validated_datapipe, NonBlockingMap):
+        return validated_datapipe
+    if not hasattr(validated_datapipe, 'nonblocking_len'):
+        def nonblocking_len(self):
+            return self.__len__()
+        validated_datapipe.nonblocking_len = types.MethodType(  # type: ignore[attr-defined]
+            nonblocking_len, validated_datapipe)
+    if not hasattr(validated_datapipe, 'nonblocking_getitem'):
+        def nonblocking_getitem(self, index):
+            return self.__getitem__(index)
+        validated_datapipe.nonblocking_getitem = types.MethodType(  # type: ignore[attr-defined]
+            nonblocking_getitem, validated_datapipe)
+    return validated_datapipe
+
+
+def DataPipeBehindQueues(source_datapipe, protocol, full_stop=False, blocking_request_get=False):
+    """
+        Indefinitely iterates over req_queue and passing values from source_datapipe to res_queue
+        If raise_stop is true, raises exception when StopIteration received from the source_datapipe
+    """
+    if not isinstance(protocol, communication.protocol.MapDataPipeQueueProtocolServer):
+        raise Exception('Expecting MapDataPipeQueueProtocolServer, got', protocol)
+    source_datapipe = EnsureNonBlockingMapDataPipe(source_datapipe)
+    forever = True
+    while forever:
+        try:
+            # Non-blocking call is Extremely slow here for python.mp, need to figure out a good workaround
+            request = protocol.get_new_request(block=blocking_request_get)
+        except communication.protocol.EmptyQueue:
+            yield True
+            continue
+
+        if isinstance(request, communication.messages.TerminateRequest):
+            forever = False
+            protocol.response_terminate()
+
+        elif isinstance(request, communication.messages.LenRequest):
+            size = source_datapipe.nonblocking_len()
+            protocol.response_len(size)
+
+        elif isinstance(request, communication.messages.GetItemRequest):
+            while forever:
+                try:
+                    value = source_datapipe.nonblocking_getitem(request.key)
+                except NotAvailable:
+                    yield True
+                    continue
+                except IndexError as e:
```

Comment:
I choose to trigger `StopIteration` when an invalid index is provided. This works well when the loop incrementally passes in indices and eventually goes out of bound. 

Do you think an alternative of raising IndexError and handling that is better?